sktime/annotation/eagglo.py

```diff
@@ -0,0 +1,430 @@
+
+        def filter_na(i):
+            return list(
+                filter(
+                    lambda v: v == v,
+                    self.progression[
+                        i,
+                    ],
+                )
+            )
+
+        # penalize the gof_ statistic
+        if self.penalty is not None:
+            penalty_func = get_penalty_func(self.penalty)
+            cps = [filter_na(i) for i in range(len(self.progression))]
+            self.gof_ += list(map(penalty_func, cps))
+
+        # get the set of change points for the "best" clustering
+        idx = np.argmax(self.gof_)
+        self._estimates = np.sort(filter_na(idx))
+
+        # remove change point N+1 if a cyclic merger was performed
+        self._estimates = (
+            self._estimates[:-1] if self._estimates[0] != 0 else self._estimates
+        )
+
+        # create final membership vector
+        def get_cluster(estimates):
+            return np.repeat(
+                range(len(np.diff(estimates))), np.diff(estimates).astype(int)
+            )
+
+        if self._estimates[0] == 0:
+            self.cluster_ = get_cluster(self._estimates)
+        else:
+            tmp = get_cluster(np.append([0], self._estimates))
+            self.cluster_ = np.append(tmp, np.zeros(X.shape[0] - len(tmp)))
+
+        return self
+
+    def _transform(self, X: pd.DataFrame, y=None):
+        """Transform X and return a transformed version.
+
+        private _transform containing core logic, called from transform
+
+        Parameters
+        ----------
+        X : Series of mtype X_inner_mtype
+            Data to be transformed
+        y : Series of mtype y_inner_mtype, default=None
+            Not required for this unsupervised transform.
+
+        Returns
+        -------
+        cluster
+            numeric representation of cluster membership for each row of X.
+        """
+        # fit again if indices not seen, but don't store anything
+        if not X.index.equals(self._X.index):
+            X_full = X.combine_first(self._X)
+            new_eagglo = EAGGLO(
+                member=self.member,
+                alpha=self.alpha,
+                penalty=self.penalty,
+            ).fit(X_full)
+            warnings.warn(
+                "Warning: Input data X differs from that given to fit(). "
+                "Refitting with both the data in fit and new input data, not storing "
+                "updated public class attributes. For this, explicitly use fit(X) or "
+                "fit_transform(X)."
+            )
+            return new_eagglo.cluster_
+
+        return self.cluster_
+
+    def _initialize_params(self, X: pd.DataFrame) -> None:
+        """Initialize parameters and store to self."""
+        self._member = np.array(
+            self.member if self.member is not None else range(X.shape[0])
+        )
+
+        unique_labels = np.sort(np.unique(self._member))
+        self.n_cluster = len(unique_labels)
```

Comment:
I find it a bit strange that `self.n_cluster` is never updated? At least from what I could find from Cntl+F of "self.n_cluster =". I would have naively expected this to be updated each time we merge 2 clusters... 🤔

Reply:
check `_update_distances` function for example